### DISCUSSION TITLE
[Feature] A command to just download and cache dependencies from lockfile

### DISCUSSION THREAD

**Discussion by rohit-gohri** [yarnpkg/berry#4380] [Ideas]
## Edit
I created a small package that recreates the full workspace structure with dummy `package.json`s just from the `yarn.lock` file - ﻿https://github.com/rohit-gohri/yarn-lock-to-package-json

Use it like this

```Dockerfile
COPY .yarn ./.yarn
COPY yarn.lock .yarnrc.yml ./

RUN yarn dlx yarn-lock-to-package-json

RUN yarn install --immutable
```

---

### Original Idea

I had an idea for something, discovered `pnpm` already has the feature. Would like to create it as a plugin, any ideas where to start? 

**Describe the user story**

As a developer using Dockerfiles with my monorepo based application, I often have to do lots of hacks to maintain docker cache of yarn install (including creating subsets of package.json with `jq`, https://stackoverflow.com/a/59606373 ). I would like to be able to separate the download and install/link steps of `yarn install` so that the download part can be cached in a separate layer (that can then be used as a base image by all the applications in the monorepo).

**Describe the solution you'd like**

A command similar to [`pnpm fetch`](https://pnpm.io/cli/fetch) that uses the  lockfile to populate the yarn cache with the dependencies required for a full install or a focus install.

This cmd will use just the `yarn.lock`, `.yarnrc.yml`, and `.yarn` folder to populate the yarn cache.

**Describe the drawbacks of your solution**

Would add some complexity and need to work without a `package.json`.

**Describe alternatives you've considered**

- Commit the cache - Too large for monorepos
- Use approach described in  https://stackoverflow.com/a/59606373 to create minimal package.json to populate the cache

**Comment by edsrzf** [comment]
I created #4529 before finding this discussion. (I only searched issues, not discussions!)

I've been able to write a plugin that does this, but ended up having to reimplement parts of the resolution process within the plugin, since there is no API to resolve dependencies without reading `package.json` files. I haven't published the code because it includes hacks that are specific to my setup. I could avoid these hacks if, for example, there were a way to configure `Project.resolveEverything` to only read `yarn.lock`.

**Comment by rohit-gohri** [comment]
I created a small package that recreates the full workspace structure with dummy `package.json`s just from the `yarn.lock` file - ﻿https://github.com/rohit-gohri/yarn-lock-to-package-json

Using it like this

```Dockerfile
COPY .yarn ./.yarn
COPY yarn.lock .yarnrc.yml ./

RUN yarn dlx yarn-lock-to-package-json

RUN yarn install --immutable
```

**Comment by devthejo** [comment]
here is my plugin (based on [yarn-lock-to-package-json](https://github.com/rohit-gohri/yarn-lock-to-package-json)):
https://github.com/devthejo/yarn-plugin-fetch
